// @zk-kit/utils/dist/lib.commonjs/conversions.cjs
/**
 * @module @zk-kit/utils
 * @version 0.6.0
 * @file Essential zero-knowledge utility library for JavaScript developers.
 * @copyright Ethereum Foundation 2024
 * @license MIT
 * @see [Github]{@link https://github.com/privacy-scaling-explorations/zk-kit/tree/main/packages/utils}
*/
'use strict';

var buffer = require('buffer');
var errorHandlers = require('./error-handlers.cjs');
var typeChecks = require('./type-checks.cjs');

/**
 * @module Conversions
 * This module provides a collection of utility functions for converting
 * between different numerical formats, particularly focusing on
 * conversions involving bigints, hexadecimals and buffers.
 * The module is structured with clear function naming to indicate
 * the conversion direction (e.g., `bigIntToHexadecimal` for BigInt
 * to hexadecimal, `bufferToBigInt` for buffer to bigint) and employs
 * type checks to ensure the correct handling of various input types.
 * It also includes variations for both big-endian (`be`) and little-endian
 * (`le`) conversions. It is important to note that when there is no prefix,
 * the order of bytes is always big-endian.
 */
/**
 * Converts a bigint to a hexadecimal string.
 * @param n The bigint value to convert.
 * @returns The hexadecimal representation of the bigint.
 */
function bigIntToHexadecimal(n) {
    let hex = n.toString(16);
    // Ensure even length.
    if (hex.length % 2 !== 0) {
        hex = `0${hex}`;
    }
    return hex;
}
/**
 * Converts a hexadecimal string to a bigint. The input is interpreted as hexadecimal
 * with or without a '0x' prefix. It uses big-endian byte order.
 * @param hex The hexadecimal string to convert.
 * @returns The bigint representation of the hexadecimal string.
 */
function beHexadecimalToBigInt(hex) {
    // Ensure the hex string starts with '0x'.
    const formattedHexString = hex.startsWith("0x") ? hex : `0x${hex}`;
    return BigInt(formattedHexString);
}
/**
 * Converts a hexadecimal string to a bigint. Alias for beHexadecimalToBigInt.
 * @param hex The hexadecimal string to convert.
 * @returns The bigint representation of the hexadecimal string.
 */
function hexadecimalToBigInt(hex) {
    return beHexadecimalToBigInt(hex);
}
/**
 * Converts a buffer of bytes to a bigint using big-endian byte order.
 * @param b The buffer to convert.
 * @returns The bigint representation of the buffer's contents.
 */
function beBufferToBigInt(b) {
    return BigInt(`0x${b.toString("hex")}`);
}
/**
 * Converts a buffer to a bigint using little-endian byte order.
 * @param buffer The buffer to convert.
 * @returns The bigint representation of the buffer's contents in little-endian.
 */
function leBufferToBigInt(buffer$1) {
    return BigInt(`0x${buffer.Buffer.from(buffer$1).reverse().toString("hex")}`);
}
/**
 * Converts a buffer to a bigint. Alias for beBufferToBigInt.
 * @param b The buffer to convert.
 * @returns The bigint representation of the buffer's contents.
 */
function bufferToBigInt(b) {
    return beBufferToBigInt(b);
}
/**
 * Converts a bigint to a buffer and fills with zeros if a valid
 * size (i.e. number of bytes) is specified. If the size is not defined,
 * it gets the size from the given bigint. If the specified size is smaller than
 * the size of the bigint (i.e. `minSize`), an error is thrown.
 * It uses big-endian byte order.
 * @param n The bigint to convert.
 * @param size The number of bytes of the buffer to return.
 * @returns The buffer representation of the bigint.
 */
function beBigIntToBuffer(n, size) {
    const hex = bigIntToHexadecimal(n);
    // Calculate the minimum buffer size required to represent 'n' in bytes.
    // Each hexadecimal character represents 4 bits, so 2 characters are 1 byte.
    const minSize = Math.ceil(hex.length / 2);
    if (!size) {
        size = minSize;
    }
    else if (size < minSize) {
        throw Error(`Size ${size} is too small, need at least ${minSize} bytes`);
    }
    // Allocate buffer of the desired size, filled with zeros.
    const buffer$1 = buffer.Buffer.alloc(size, 0);
    const fromHex = buffer.Buffer.from(hex, "hex");
    fromHex.copy(buffer$1, size - fromHex.length);
    return buffer$1;
}
/**
 * Converts a bigint to a buffer and fills with zeros if a valid
 * size (i.e. number of bytes) is specified. If the size is not defined,
 * it gets the size from the given bigint. If the specified size is smaller than
 * the size of the bigint (i.e. `minSize`), an error is thrown.
 * It uses little-endian byte order.
 * @param n The bigint to convert.
 * @param size The number of bytes of the buffer to return.
 * @returns The buffer representation of the bigint in little-endian.
 */
function leBigIntToBuffer(n, size) {
    const hex = bigIntToHexadecimal(n);
    // Calculate the minimum buffer size required to represent 'n' in bytes.
    // Each hexadecimal character represents 4 bits, so 2 characters are 1 byte.
    const minSize = Math.ceil(hex.length / 2);
    if (!size) {
        size = minSize;
    }
    else if (size < minSize) {
        throw Error(`Size ${size} is too small, need at least ${minSize} bytes`);
    }
    // Allocate buffer of the desired size, filled with zeros.
    const buffer$1 = buffer.Buffer.alloc(size, 0);
    const fromHex = buffer.Buffer.from(hex, "hex").reverse();
    fromHex.copy(buffer$1, 0);
    return buffer$1;
}
/**
 * Converts a bigint to a buffer. Alias for beBigIntToBuffer.
 * @param n The bigint to convert.
 * @returns The buffer representation of the bigint.
 */
function bigIntToBuffer(n) {
    return beBigIntToBuffer(n);
}
/**
 * Converts a BigNumberish type to a bigint. If the input is already a bigint,
 * the return value will be the bigint itself, otherwise it will be converted
 * to a bigint using big-endian byte order.
 * @param n The BigNumberish value to convert.
 * @returns The bigint representation of the BigNumberish value.
 */
function bigNumberishToBigInt(n) {
    if (typeChecks.isNumber(n) || typeChecks.isBigInt(n) || typeChecks.isStringifiedBigInt(n) || typeChecks.isHexadecimal(n)) {
        return BigInt(n);
    }
    return bufferToBigInt(n);
}
/**
 * Converts a BigNumberish type to a buffer. If the input is already a buffer,
 * the return value will be the buffer itself, otherwise it will be converted
 * to a buffer using big-endian byte order.
 * @param n The BigNumberish value to convert.
 * @returns The buffer representation of the BigNumberish value.
 */
function bigNumberishToBuffer(n) {
    if (typeChecks.isBuffer(n)) {
        return n;
    }
    return bigIntToBuffer(bigNumberishToBigInt(n));
}
/**
 * Converts an hexadecimal string to a buffer. The hexadecimal string
 * should not start with '0x' or '0X'. It keeps the bytes in the same order.
 * @param value The hexadecimal string to convert.
 * @returns The buffer representation of the hexadecimal string.
 */
function hexadecimalToBuffer(value) {
    errorHandlers.requireHexadecimal(value, "value", false);
    return buffer.Buffer.from(value, "hex");
}
/**
 * Converts a buffer to a hexadecimal string. It accepts 'Buffer' or 'Uint8Array'.
 * The hexadecimal string will not start with '0x' or '0X'. It keeps the bytes in the same order.
 * @param value The buffer to convert.
 * @returns The converted hexadecimal string.
 */
function bufferToHexadecimal(value) {
    errorHandlers.requireTypes(value, "value", ["Buffer", "Uint8Array"]);
    return buffer.Buffer.from(value).toString("hex");
}

exports.beBigIntToBuffer = beBigIntToBuffer;
exports.beBufferToBigInt = beBufferToBigInt;
exports.beHexadecimalToBigInt = beHexadecimalToBigInt;
exports.bigIntToBuffer = bigIntToBuffer;
exports.bigIntToHexadecimal = bigIntToHexadecimal;
exports.bigNumberishToBigInt = bigNumberishToBigInt;
exports.bigNumberishToBuffer = bigNumberishToBuffer;
exports.bufferToBigInt = bufferToBigInt;
exports.bufferToHexadecimal = bufferToHexadecimal;
exports.hexadecimalToBigInt = hexadecimalToBigInt;
exports.hexadecimalToBuffer = hexadecimalToBuffer;
exports.leBigIntToBuffer = leBigIntToBuffer;
exports.leBufferToBigInt = leBufferToBigInt;
